integration-tests/checkstyle-10.12.4.sh

```diff
@@ -0,0 +1,162 @@
+git commit -m 'dependency: Introduce Error Prone Support' .
+mvn ${shared_build_flags} 'com.spotify.fmt:fmt-maven-plugin:${fmt.version}:format'
+git commit -m 'minor: Reformat using Google Java Format' .
+diff_base="$(git rev-parse HEAD)"
+
+# Apply Error Prone Support-suggested changes until a fixed point is reached.
+function apply_patch() {
+  local extra_build_args="${1}"
+
+  mvn ${shared_build_flags} ${extra_build_args} \
+    package 'com.spotify.fmt:fmt-maven-plugin:${fmt.version}:format' \
+    -Derror-prone.flags="${error_prone_patch_flags}" \
+    -DskipTests
+
+  if ! git diff --exit-code; then
+    git commit -m 'minor: Apply patches' .
+
+    # Changes were applied, so another compilation round may apply yet more
+    # changes. For performance reasons we perform incremental compilation,
+    # enabled using a misleading flag. (See
+    # https://issues.apache.org/jira/browse/MCOMPILER-209 for details.)
+    apply_patch '-Dmaven.compiler.useIncrementalCompilation=false'
+  elif [ "${extra_build_args}" != 'clean' ]; then
+    # No changes were applied. We'll attempt one more round in which all files
+    # are recompiled, because there are cases in which violations are missed
+    # during incremental compilation.
+    apply_patch 'clean'
+  fi
+}
+apply_patch ''
+
+# Perist or validate the applied changes.
+baseline_patch="${integration_test_root}/${test_name}-expected-changes.patch"
+if [ -n "${do_sync}" ]; then
+  echo 'Saving changes...'
+  git diff "${diff_base}" | grep -vP '^(diff|index)' > "${baseline_patch}"
+else
+  echo 'Inspecting changes...'
+  # XXX: This "diff of diffs" also contains vacuous sections, introduced due to
+  # line offset differences. Try to omit those from the final output.
+  if ! diff -u "${baseline_patch}" <(git diff "${diff_base}" | grep -vP '^(diff|index)'); then
+    echo 'There are unexpected changes.'
+    exit 1
+  fi
+fi
+
+# Run one more full build and log the output.
+#
+# By also running the tests, we validate that the (majority of) applied changes
+# are behavior preserving. Some tests are skipped:
+# - The `metadataFilesGenerationAllFiles` test is skipped because it makes line
+#   number assertions that will fail when the code is formatted or patched.
+# - The `allCheckSectionJavaDocs` test is skipped because is validates that
+#   Javadoc has certain closing tags that are removed by Google Java Format.
+mvn ${shared_build_flags} \
+    clean package \
+    -Derror-prone.flags="${error_prone_validation_flags}" \
+    -Dtest='
+      !MetadataGeneratorUtilTest#metadataFilesGenerationAllFiles,
+      !XdocsJavaDocsTest#allCheckSectionJavaDocs' \
+  | tee "${validation_log_file}"
+
+# Perist or validate the warnings reported by Error Prone Support checks.
+baseline_warnings="${integration_test_root}/${test_name}-expected-warnings.txt"
+generated_warnings="$((grep -oP "(?<=^\\Q[WARNING] ${PWD}/\\E).*" "${validation_log_file}" | grep -P '\] \[' || true) | sort)"
```

Comment:
Here and elsewhere, the use of `grep -P` likely doesn't make this script Macos-compatible; that's something to review separately.